Rust generator emits conflicting module and struct names for messages with nested types

### Search before asking

- [x] I had searched in the [issues](https://github.com/apache/fory/issues) and found no similar issues.


### Version

0.16.0

### Component(s)

Rust

### Minimal reproduce step

FDL:
```
message foo {
    message Bar {
        string baz = 1;
    }
    Bar bar = 1;
}
```
Generated Rust code:
```rust
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.
//
// This file is generated by Apache Fory compiler.

use fory::{Fory, ForyObject};
use std::sync::OnceLock;

pub mod foo {
    use super::*;

    #[derive(ForyObject, Debug, Clone, PartialEq, Default)]
    pub struct Bar {
        #[fory(id = 1)]
        pub baz: String,
    }

    impl Bar {
        pub fn to_bytes(&self) -> Result<Vec<u8>, fory::Error> {
            let fory = detail::get_fory();
            fory.serialize(self)
        }

        pub fn from_bytes(data: &[u8]) -> Result<Bar, fory::Error> {
            let fory = detail::get_fory();
            fory.deserialize(data)
        }
    }
}

#[derive(ForyObject, Debug, Clone, PartialEq, Default)]
pub struct foo {
    #[fory(id = 1)]
    pub bar: foo::Bar,
}

impl foo {
    pub fn to_bytes(&self) -> Result<Vec<u8>, fory::Error> {
        let fory = detail::get_fory();
        fory.serialize(self)
    }

    pub fn from_bytes(data: &[u8]) -> Result<foo, fory::Error> {
        let fory = detail::get_fory();
        fory.deserialize(data)
    }
}

pub fn register_types(fory: &mut Fory) -> Result<(), fory::Error> {
    fory.register_by_namespace::<foo::Bar>("default", "foo.Bar")?;
    fory.register_by_namespace::<foo>("default", "foo")?;
    Ok(())
}

mod detail {
    use super::*;

    pub(super) fn get_fory() -> &'static Fory {
        static FORY: OnceLock<Fory> = OnceLock::new();
        FORY.get_or_init(|| {
            let mut fory = Fory::default()
                .xlang(true)
                .track_ref(true)
                .compatible(true);
            register_types(&mut fory).expect("failed to register fory types");
            fory
        })
    }
}
```

### What did you expect to see?

The generated Rust code should compile.

### What did you see instead?

Compiling errors:
```
    Checking old_rust_escape_check v0.1.0 (/tmp/fory-old-rust-compile-sM2lgt)
error[E0428]: the name `foo` is defined multiple times
  --> src/generated.rs:46:1
   |
23 | pub mod foo {
   | ----------- previous definition of the module `foo` here
...
46 | pub struct foo {
   | ^^^^^^^^^^^^^^ `foo` redefined here
   |
   = note: `foo` must be defined only once in the type namespace of this module

error[E0573]: expected type, found module `foo`
  --> src/generated.rs:46:12
   |
46 | pub struct foo {
   |            ^^^ not a type

error[E0573]: expected type, found module `foo`
  --> src/generated.rs:45:22
   |
45 | #[derive(ForyObject, Debug, Clone, PartialEq, Default)]
   |                      ^^^^^ not a type

error[E0573]: expected type, found module `foo`
  --> src/generated.rs:45:29
   |
45 | #[derive(ForyObject, Debug, Clone, PartialEq, Default)]
   |                             ^^^^^ not a type

error[E0574]: expected struct, variant or union type, found module `foo`
  --> src/generated.rs:45:29
   |
45 | #[derive(ForyObject, Debug, Clone, PartialEq, Default)]
   |                             ^^^^^ not a struct, variant or union type

error[E0573]: expected type, found module `foo`
  --> src/generated.rs:45:36
   |
45 | #[derive(ForyObject, Debug, Clone, PartialEq, Default)]
   |                                    ^^^^^^^^^ not a type

error[E0573]: expected type, found module `foo`
  --> src/generated.rs:45:47
   |
45 | #[derive(ForyObject, Debug, Clone, PartialEq, Default)]
   |                                               ^^^^^^^ not a type

error[E0574]: expected struct, variant or union type, found module `foo`
  --> src/generated.rs:45:47
   |
45 | #[derive(ForyObject, Debug, Clone, PartialEq, Default)]
   |                                               ^^^^^^^ not a struct, variant or union type

error[E0573]: expected type, found module `foo`
  --> src/generated.rs:51:6
   |
51 | impl foo {
   |      ^^^ not a type

error[E0573]: expected type, found module `foo`
  --> src/generated.rs:57:46
   |
57 |     pub fn from_bytes(data: &[u8]) -> Result<foo, fory::Error> {
   |                                              ^^^ not a type

error[E0573]: expected type, found module `foo`
  --> src/generated.rs:65:34
   |
65 |     fory.register_by_namespace::<foo>("default", "foo")?;
   |                                  ^^^ not a type

warning: type `foo` should have an upper camel case name
  --> src/generated.rs:46:12
   |
46 | pub struct foo {
   |            ^^^ help: convert the identifier to upper camel case (notice the capitalization): `Foo`
   |
   = note: `#[warn(non_camel_case_types)]` (part of `#[warn(nonstandard_style)]`) on by default

Some errors have detailed explanations: E0428, E0573, E0574.
For more information about an error, try `rustc --explain E0428`.
warning: `old_rust_escape_check` (lib) generated 1 warning
error: could not compile `old_rust_escape_check` (lib) due to 11 previous errors; 1 warning emitted

```

### Anything Else?

_No response_

### Are you willing to submit a PR?

- [x] I'm willing to submit a PR!

--- a/compiler/fory_compiler/generators/rust.py
+++ b/compiler/fory_compiler/generators/rust.py
@@ -172,9 +172,9 @@ class RustGenerator(BaseGenerator):
         if "." in type_name:
             parts = type_name.split(".")
             parents = [self.to_snake_case(name) for name in parts[:-1]]
-            path = "::".join(parents + [parts[-1]])
+            path = "::".join(parents + [self.to_pascal_case(parts[-1])])
             return f"crate::{module}::{path}"
-        return f"crate::{module}::{type_name}"
+        return f"crate::{module}::{self.to_pascal_case(type_name)}"
 
     def generate_bytes_impl(self, type_name: str) -> List[str]:
         lines = []
@@ -413,7 +413,7 @@ class RustGenerator(BaseGenerator):
         """Generate a Rust struct."""
         lines = []
 
-        type_name = message.name
+        type_name = self.to_pascal_case(message.name)
 
         # Derive macros
         comment = self.format_type_id_comment(message, "//")
@@ -487,7 +487,7 @@ class RustGenerator(BaseGenerator):
     def generate_debug_impl(self, message: Message) -> List[str]:
         """Generate a Debug impl that avoids recursive ref expansion."""
         lines: List[str] = []
-        type_name = message.name
+        type_name = self.to_pascal_case(message.name)
         lines.append(f"impl std::fmt::Debug for {type_name} {{")
         lines.append(
             "    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {"
@@ -766,20 +766,20 @@ class RustGenerator(BaseGenerator):
             target_parents = parts[:-1]
             base_name = parts[-1]
             return self.build_relative_type_name(
-                current_parents, target_parents, base_name
+                current_parents, target_parents, self.to_pascal_case(base_name)
             )
         if not parent_stack:
-            return type_name
+            return self.to_pascal_case(type_name)
 
         for i in range(len(parent_stack) - 1, -1, -1):
             message = parent_stack[i]
             if message.get_nested_type(type_name) is not None:
                 target_parents = [msg.name for msg in parent_stack[: i + 1]]
                 return self.build_relative_type_name(
-                    current_parents, target_parents, type_name
+                    current_parents, target_parents, self.to_pascal_case(type_name)
                 )
 
-        return type_name
+        return self.to_pascal_case(type_name)
 
     def collect_uses(self, field_type: FieldType, uses: Set[str]):
         """Collect required use statements for a field type."""
@@ -921,7 +921,7 @@ class RustGenerator(BaseGenerator):
         parent_stack: Optional[List[Message]],
     ):
         """Generate registration code for a message and its nested types."""
-        type_name = self.get_type_path(message.name, parent_stack)
+        type_name = self.get_type_path(self.to_pascal_case(message.name), parent_stack)
         reg_name = self.get_registration_type_name(message.name, parent_stack)
 
         # Register nested enums first

--- a/compiler/fory_compiler/tests/test_nested_types.py
+++ b/compiler/fory_compiler/tests/test_nested_types.py
@@ -17,10 +17,15 @@
 
 """Tests for FDL nested type support."""
 
+from pathlib import Path
+from textwrap import dedent
+
 import pytest
 
 from fory_compiler.frontend.fdl.lexer import Lexer
 from fory_compiler.frontend.fdl.parser import Parser
+from fory_compiler.generators.base import GeneratorOptions
+from fory_compiler.generators.rust import RustGenerator
 from fory_compiler.ir.ast import NamedType, ListType
 from fory_compiler.ir.validator import SchemaValidator
 
@@ -367,6 +372,77 @@ class TestSchemaTypeLookup:
         assert "Status" in type_names
         assert "Inner" in type_names
         assert "Deep" in type_names
+
+
+class TestRustNestedTypeGeneration:
+    """Rust-specific tests for nested type code generation."""
+
+    def _generate_rust(self, source: str) -> str:
+        lexer = Lexer(source)
+        parser = Parser(lexer.tokenize())
+        schema = parser.parse()
+        options = GeneratorOptions(output_dir=Path("/tmp"))
+        generator = RustGenerator(schema, options)
+        files = generator.generate()
+        return "\n".join(f.content for f in files)
+
+    def test_nested_message_no_name_collision(self):
+        """Regression for #3547: struct and module must not share the same identifier."""
+        source = dedent("""
+            message foo {
+                message Bar {
+                    string baz = 1;
+                }
+                Bar bar = 1;
+            }
+        """)
+        content = self._generate_rust(source)
+        assert "pub mod foo {" in content
+        assert "pub struct Foo {" in content
+        assert "pub struct foo" not in content
+
+    def test_nested_message_field_type_qualified(self):
+        """Field referencing a nested type resolves to module-qualified path."""
+        source = dedent("""
+            message foo {
+                message Bar {
+                    string baz = 1;
+                }
+                Bar bar = 1;
+            }
+        """)
+        content = self._generate_rust(source)
+        assert "pub bar: foo::Bar," in content
+
+    def test_nested_message_registration_uses_pascal_struct(self):
+        """register_by_namespace must reference the PascalCase struct name."""
+        source = dedent("""
+            message foo {
+                message Bar {
+                    string baz = 1;
+                }
+                Bar bar = 1;
+            }
+        """)
+        content = self._generate_rust(source)
+        assert "register_by_namespace::<Foo>" in content
+        assert 'register_by_namespace::<Foo>("default", "foo")' in content
+        assert "register_by_namespace::<foo::Bar>" in content
+
+    def test_pascal_case_names_unchanged(self):
+        """Messages already in PascalCase must not be renamed."""
+        source = dedent("""
+            message Outer {
+                message Inner {
+                    string value = 1;
+                }
+                Inner inner = 1;
+            }
+        """)
+        content = self._generate_rust(source)
+        assert "pub struct Outer {" in content
+        assert "pub struct Inner {" in content
+        assert "pub outer: " not in content
 
 
 if __name__ == "__main__":

